Implement alpha version of PreferAvoidPods

jiangyaoguo opened the pull request:
This is part of #18853

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/20699)

<!-- Reviewable:end -->

davidopp commented:
cc/ @smarterclayton @mikedanese @bgrant0607

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -2229,6 +2229,11 @@ type DeleteOptions struct {
 	// Defaults to a per object value if not specified. zero means delete immediately.
 	GracePeriodSeconds *int64 `json:"gracePeriodSeconds"`
+
+	// If true, when deleting the pod managed by a replication controller, scheduler will try to put the replacement
```
Slightly different wording that might be a bit clearer: "If the object to be deleted is a pod managed by a controller, setting to true requests that the replacement pod that the controller creates (if the controller decides to create such a pod) be scheduled to a different node than the one where it is currently running. Otherwise has no effect."

smarterclayton commented:
I thought we said in the other thread this needs to be on a new API
object and that we have to be able to convert both DeleteOptions and
PodDeleteOptions to the internal type?

davidopp commented:
Here are the two comments I think you're referring to:

https://github.com/kubernetes/kubernetes/issues/18853#issuecomment-168799279
https://github.com/kubernetes/kubernetes/issues/18853#issuecomment-168800195

I interpreted these comments to mean that anything beyond what @jiangyaoguo was a "nice to have."

Do we foresee other Pod-specific delete options? It doesn't seem so bad to just put in a comment "has no effect when applied to non-Pod" unless there are going to be multiple other scenarios like this.

cc/ @lavalamp

smarterclayton commented:
We can always change the schema in the future to take away the flag since
it would have no effect, but we wouldn't be able to be stricter in
validating incoming requests, because the field would be accepted by
version 1.2 (even if ignored).  So it paints us into a corner a little bit
more than today.

On Sun, Feb 7, 2016 at 2:55 PM, David Oppenheimer <notifications@github.com>
wrote:

> Here are the two comments I think you're referring to:
>
> #18853 (comment)
> <https://github.com/kubernetes/kubernetes/issues/18853#issuecomment-168799279>
> #18853 (comment)
> <https://github.com/kubernetes/kubernetes/issues/18853#issuecomment-168800195>
>
> I interpreted these comments to mean that anything beyond what
> @jiangyaoguo <https://github.com/jiangyaoguo> was a "nice to have."
>
> Do we foresee other Pod-specific delete options? It doesn't seem so bad to
> just put in a comment "has no effect when applied to non-Pod" unless there
> are going to be multiple other scenarios like this.
>
> cc/ @lavalamp <https://github.com/lavalamp>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20699#issuecomment-181094241>
> .
>

jiangyaoguo commented:
@k8s-bot test this please

lavalamp commented:
I would really prefer to make a specific PodDeletionOptions object. How urgent is this?

davidopp commented:
Quick notes from conversation with @lavalamp today:
* The fact that AvoidPreviousNode was requested in the api.DeleteOptions needs to be persisted somewhere by the API server, presumably an annotation on the pod that it then deletes. An alternative would be to just have the client who requests the delete add the annotation before doing the delete, and not bother with adding this option to DeleteOptions at all.
* More controversial alternatives
  * have a new API resource that records the information; it would be available to all controllers (though it's not clear that other controllers than the one that created the pod need the information), and could eventually generalize to a RateLimiter-type object that records pod-machine pairs to avoid
  * similar to previous sub-bullet but put the information in the Node object rather than a new object

I'm pretty opposed to the "more controversial alternatives." It seems overly complicated, at least in the sense that associating  the "avoid previous machine" information with the pod that is being deleted avoids any concern about garbage collecting the information, as compared to storing it in some global structure (as in either of the two "more controversial alternatives").

I'm on the fence about the first bullet.

Urgency: This is not super-urgent; it blocks deployment of rescheduler, but we haven't even started implementing rescheduler.

jiangyaoguo commented:
@k8s-bot test this please

jiangyaoguo commented:
Agree with @davidopp . It's overly complicated to add a separate API object and make DeleteOptions be polymorphic. 
@smarterclayton  what kinds of validation will be added to kubernetes? It seems current validation has no block.

davidopp commented:
@jiangyaoguo I think the most expedient thing to do would be as @lavalamp suggested, just use an annotation on the pod rather than changing api.DeleteOptions. It makes the interface a little more ugly for clients -- they would have to add the annotation and then delete the object, rather than just deleting the object with the proper deletion options -- but it avoids the debate whether we need to create a new kind of delete options. If we do this, then we don't need this PR.

But, we would still need a PR that implements the rest of the feature--
1) define the name of the annotation key (like "scheduler.alpha.kubernetes.io/avoidPreviousNode") as a constant somewhere (only the annotation's key is used, there is no value)
2) define a second annotation which will be used to tell the scheduler which node to avoid putting a pod onto (the key is something like "scheduler.alpha.kubernetes.io/avoidNode", and the value is the name of the node to avoid)
3) modify the controllers so that when they see a pod get deleted, if the pod had the annotation 1 and they decide to create a replacement pod, they add the annotation 2 when they create the replacement pod. This requires the controller to "remember" every time it sees annotation 1 added to a pod, so that when it creates a replacement pod it knows to add annotation 2 and what value to use (the hostname that the deleted pod was on). This "remembering" does not have to be persisted; if the controller fails between annotation 1 added and replacement pod being created, it is OK that replacement pod will not have annotation 2. This feature is "best effort" and controller failure should be rare.
4) modify the scheduler so that it always avoids putting the replacement pod on the node indicated by the annotation 2 value, *unless* that node is the only one that meets all the predicates.

jiangyaoguo commented:
A deletion of pod will cause RC-controller to sync RC of deleted pod to its expected status. But the  replacement pod is not strictly correlative to the deleted pod. About feature 3), after deletion of a pod, what we really need is one of following pods that be created by RC should avoid this previous node, or all of  following pods should avoid?  How long should controller keep creating pod that avoid previous node?  The former seems more clear logically. 

If we have added annotation 1 to pod,   it means the "remembering" is persisted. Controller can get this information directly when it see a pod get deleted. We don't need extra "remembering".

lavalamp commented:
Just to continue playing devil's advocate, setting an annotation on the pod (or adding a field) gives schedulers only the deletion period to notice the change--essentially, you're forcing schedulers/controllers to start keeping state; currently, they're stateless. (Of course, in the future we'll be storing deleted pods for some amount of time, but currently we do not, and it's a fairly big change, so you may not want to block this on waiting for that.)

To me, the old pod object still seems like a weird place to store this information. I still think you should consider a separate resource entirely, where you can record pod / node incompatibilities. You could additionally store things like a duration for the incompatibility, or other condition that would cause the incompatibility to expire. This solution means that we don't have to store local state in the schedulers/controllers. I also like this solution because it might make you figure out what exactly about pod X implies that pod Y should not go on the node--that is a huge mystery to me given what you're currently talking about doing.

Or you could be even more general, and start a rate limiter type service, but I think that's overly general given what you're trying to solve.

davidopp commented:
> Just to continue playing devil's advocate, setting an annotation on the pod (or adding a field) gives schedulers only the deletion period to notice the change--essentially, you're forcing schedulers/controllers to start keeping state; currently, they're stateless. 

I don't understand. The only requirement here is that when a controller is creating a replacement pod, it remembers whether the pod it is replacing had annotation 1 on it, and if so, what machine the pod was running on. And this is best-effort so there is no expectation that the "avoid" will be respected if the controller crashes between annotation 1 being added and the controller creating the replacement pod. So yes, the controller must store state, but only in memory.

The main use case for this feature is rescheduler. Rescheduler wants to move a pod off of a machine to make room for some other pod.

The thing I don't like about having a separate object (other than it being more work) is then you need some garbage collection mechanism. By recording the "avoid previous node" bit in the deleted pod, the garbage collection is solved.

However, if we think we're going to need a new API object to record "avoid machine" for other purposes, then I agree we might as well bite the bullet and implement it now and use it for this purpose.

lavalamp commented:
> The only requirement here is that when a controller is creating a replacement pod, it remembers whether the pod it is replacing had annotation 1 on it, and if so, what machine the pod was running on.

I guess you're saying it's 1-shot? Only a single pod is rejected? I don't think that really makes sense from the controller's point of view. Imagine two pods die about the same time, one from rescheduler, one by the node dying (say). The first one it properly causes avoid the forbidden node, the second one it doesn't.

And the way it requires state, is that the controller has to remember this annotation thingy. The pod deletion and creation handlers in the controller are not synchronous IIRC. Some amount of time may elapse between the two events. Also there's cases where it will reboot and miss an update, or even if it's just temporarily lost the connection and is relisting, it could miss the final state of an object.

Also, I'm not entirely sure how you'll implement that with our current RC controller. Are you planning on adding a node selector to the new pod or something?

You say it's best effort; if that means "works 50% of the time" then I'll probably shut up. If that means 90% of the time, maybe I should keep arguing.

davidopp commented:
I discussed this a bit with @lavalamp today.

I think he somewhat persuaded me that we should use an object (some new API object we'd create) rather than an annotation on the to-be-deleted pod. It would be a rate limiter object in a sense. One implementation is as a map from controller to map from node name to either "avoid" or "prefer." The semantics are that when a scheduler schedules a pod created by the corresponding controller, it should avoid or prefer the indicated node. The way we'd use this with rescheduler is that when rescheduler wants to evict a pod managed by controller C from node N in order to make room for a pod managed by controller D, it would add two entries to the object:
C, {N, "avoid"}
D, {N, "prefer"}
and then delete C and D.

(There are some details about how you represent in-use resources during graceful deletion that are TBD, but that's outside the scope of this issue.)

Schedulers would interpret this to mean "when scheduling a pod owned by controller C, avoid putting it on node N unless no other nodes are feasible, and when scheduling a pod owned by controller D, prefer putting it on node N unless no other nodes are feasible."

The exact way we'd identify controllers (C and D) is TBD; probably the best approach is however we do controllerRef, but label selector might also work ("pod matching this label selector" rather than "pod owned by controller X").

Whether the outer map key should be controller or node is also somewhat unclear; for now I just randomly picked controller. One argument in favor of controller is that we'd want the indicated controller to be responsible for garbage collecting the entries that list it, and if controller is the map key then it's easy for it to find all of its pods. Another argument in favor of controller is that if we want this object to be namespace-scoped rather than a singleton for the whole cluster, you need the outer map key to be the controller (which is namespace-scoped) and not the node.

This object could also be used by a control system that notices when some controller's pods are repeatedly failing on some node (e.g. due to some bad interaction with the node) and adds an entry to temporarily prevent it from scheduling there. One example is the "kubelet admission controller knows something the scheduler doesn't" scenario; you can detect the scheduling loop and add an entry to this object in order to avoid a tight rescheduling loop onto the same machine where the pod keeps getting rejected by the kubelet.The are other more esoteric possibilities like a pod that for whatever reason deterministically causes a machine to crash, or gets poor performance on a particular machine, or whatever.

We could possibly have a kubectl command that allows user to add entries to the object, but I'm not sure about the security/safety implications of that.

A few of the benefits of this approach over the one I was advocating (I think @lavalamp already mentioned all of these)
* more debuggable--because these "avoid" hints are in an easily-discoverable place rather than in a pod that's about to be deleted and then in a controller's in-memory state
* survives controller reboots
* can be used for non-rescheduler use cases, like the ones I mentioned earlier

Borg has something very similar to this (it's called the rate limiter.)

davidopp commented:
@mikedanese suggested (and actually @lavalamp had also suggested, but I forgot) that we could put this information into the NodeSpec or NodeStatus, and thereby avoid having a separate API object. I like this idea.

NodeStatus seems more correct to me, but since Unschedulable is in Spec, and this is pretty similar to Unschedulable, for consistency we should probably put it in Spec.

So we'd add a field to NodeSpec like
```
RateLimiter *RateLimiter
```

defined as

```
// @bgrant0607 said this is the right subset of ObjectReference to use for ControllerRef 
type ControllerRef struct {
   Kind string
   Namespace string
   Name string 
   APIVersion string   // for API group   
}

type RateLimiterRuleType string

const (
   RateLimiterRuleAvoidNode RateLimiterRuleType = "avoidNode"
   RateLimiterRulePreferNode RateLimiterRuleType = "preferNode"
)

type RateLimiterEntry {
   Controller ControllerRef
   RuleType RateLimiterRuleType
}

type RateLimiter struct {
    RateLimiterEntries []RateLimiterEntry
}
```

bgrant0607 commented:
Is this a 1.2 feature?

bgrant0607 commented:
FWIW, I don't think we should allow cross-namespace control, so controllerRef doesn't strictly need namespace.

bgrant0607 commented:
A few drive-by comments:
* Eventually I'd like to keep objects around for a bounded amount of time after they've been deleted. What Borg does is index objects by UID and use name as a symlink. On deletion, the name symlink is removed. This would allow final state to be inspected, would maintain access to logs, would facilitate server-side cascading deletion, etc.
* I'd like a reason parameter on delete, which would be visible as part of the final resource state: #1462
* We may also want to avoid scheduling on the same node in cases other than deletion, such as in the case of failure due to OOM. 
* A Kubelet-initiated deletion case: #18724
* This is a at-the-moment hint, not permanent desired state. It's unclear when the hint should be invalidated due to changing conditions on the node. Would there be a TTL? Is that part of the RateLimiter proposal? 
* I wouldn't overtrain on the "prefer" case. We need to do that in Borg in large part due to its preemption model.
* unschedulable is in spec so that it can be set by administrators, independent of conditions reported by the node that might make it unschedulable. I want to replace that with taints at some point.
* we're unhappy with NodeStatus being written by multiple entities already, or at least the same fields of NodeStatus being written by multiple entities
* We also need a representation of disruption budget (#12611), which is another potential place where this information could be recorded
* Unbounded lists have been problems in API objects in the past (e.g., BoundPods, Endpoints), and I try to avoid that in the API wherever possible. It's not scalable. This is one reason why I like the idea of annotating pods.
* Another possible approach would be for the rescheduler to use taints and tolerations to coerce pods to land where it wants them. For instance, it could put a `reserve-for-big-pod` taint on some node where it wants to squat on capacity for a pod requiring lots of resources to land there.

bgrant0607 commented:
Yet another option is to create another polymorphic controller subresource similar to /scale, say /move (which is also conveniently both a noun and verb), which could be invoked on the responsible controller identified by `controllerRef` to request the pod in question to be moved gracefully by the controller. That controller could generate a one-time anti-affinity constraint on the replacement pod.

The rolling restart request is somewhat related: #13488.

davidopp commented:
> Eventually I'd like to keep objects around for a bounded amount of time after they've been deleted. What Borg does is index objects by UID and use name as a symlink. On deletion, the name symlink is removed. This would allow final state to be inspected, would maintain access to logs, would facilitate server-side cascading deletion, etc.

The thing we need to timeout here isn't an object, but an entry in a map (either a map in a rate limiter object or a new field in each node). Oh, maybe you were talking about the original proposal of putting the info as an annotation of the to-be-deleted pod. @lavalamp convinced me to drop that proposal.

> I'd like a reason parameter on delete, which would be visible as part of the final resource state: #1462

Yes, I agree this is a good idea, and we can definitely add that to api.DeleteOptions, especially since it can apply to every object, so it doesn't have the issue of being pod-specific that @lavalamp and @smarterclayton were concerned about. But @lavalamp convinced me we shouldn't use DeleteOptions for the rescheduling stuff.

> We may also want to avoid scheduling on the same node in cases other than deletion, such as in the case of failure due to OOM. A Kubelet-initiated deletion case: #18724

Yes. Also: pod of death, and Kubelet admission control rejection loop (when Kubelet "knows more" than the scheduler).

> This is a at-the-moment hint, not permanent desired state. It's unclear when the hint should be invalidated due to changing conditions on the node.

We can make this arbitrarily complicated. My hope is to find something simple; the null hypothesis is that the Borg mechanism (plus a "prefer parent" analogue) is sufficient, but we should of course also consider alternatives if they are better from a complexity-effectiveness standpoint. For the rescheduler case (evict pod X from node M so that already-running-somewhere-else pod Y can schedule where X was), what you *really* want is to atomically replace X with Y. There are a number of reasons this doesn't work (not the least of which is that we're really talking about replacing X with the pod that Y's controller creates when Y is deleted, which is not Y). A more realizable alternative is evict X from M, atomically replace it with a balloon of size sizeof(Y) on M, then evict Y from wherever it is running, then when Y's controller creates its replacement atomically swap it with the balloon. But this balloon business is complicated. So the next level of simplification is to drop the balloon and just do the evictions, with a hint that Y's replacement should go on M and X's replacement should not go on M, which is what's being proposed here.

>  Would there be a TTL? Is that part of the RateLimiter proposal?

In this proposal it is up to the controller to decide when to remove the entry from the map. The protos I wrote above suggest that whoever adds the entry can suggest a timeout period, but it would just be a hint.

> I wouldn't overtrain on the "prefer" case. We need to do that in Borg in large part due to its preemption model.

The reason for the "prefer" here is to make it likely that Y will go to M instead of somewhere else. In the Borg rescheduler when we are just trying to get a pending pod to schedule, we don't care where it goes as long as it goes somewhere. But the anticipated use cases for the Kubernetes rescheduler include moving pods to a specific machine (or perhaps in theory a set of machines). So I think prefer is needed.

> unschedulable is in spec so that it can be set by administrators, independent of conditions reported by the node that might make it unschedulable. I want to replace that with taints at some point.

Good point. So that's not a good reason to put this information in the spec as opposed to the status.

> we're unhappy with NodeStatus being written by multiple entities already, or at least the same fields of NodeStatus being written by multiple entities

Yeah, @lavalamp brought this up too. But this would be a new field so not quite as bad as what we're doing with NodeCondition. 

> We also need a representation of disruption budget (#12611), which is another potential place where this information could be recorded

Yes, we could use the same object for both, though I think we'd end up just having two very different things jammed into the same object. The object we're talking about here is a map from controller to node (or vice-versa) whereas disruption budget is something like a map from controller to remaining budget. 

> Unbounded lists have been problems in API objects in the past (e.g., BoundPods, Endpoints), and I try to avoid that in the API wherever possible. It's not scalable. This is one reason why I like the idea of annotating pods.

The concern about unbounded lists is reasonable, and I agree that one of the big benefits of attaching the information to the pods is that you get automatic garbage collection (as well as dispersing the amount of data into smaller objects). However, it seems that we can keep the data from becoming unbounded by garbage collecting it, which we need to do anyway. For example, we can treat the object (a global object or per-node) as a fixed-size FIFO in addition to having controllers be responsible for deleting entries that talk about them after some amount of time. So you could put a ard bound on the size.

> Another possible approach would be for the rescheduler to use taints and tolerations to coerce pods to land where it wants them. For instance, it could put a reserve-for-big-pod taint on some node where it wants to squat on capacity for a pod requiring lots of resources to land there.

I couldn't tell which of these things you're suggesting:

(1) Something that would be set up ahead of time, not with the specific goal of moving a particular pod from one particular machine to another particular machine, but to implement a general policy. I can see this maybe being useful for defragmentation but I'm not sure it can handle all of the rescheduler use cases we have. But I haven't thought about it much. Also I'm not sure exactly how you decide where/when to set up these taints. If you think this idea is worth pursuing, it would be good to provide some specific examples, that cover the different use cases.

or

(2) A mechanism to implement the balloon concept I mentioned earlier in this response. @lavalamp and I actually discussed this possibility briefly, the main problem is that the most straightforward approach to using taints would block all pods (except the one you're trying to move there) from scheduling onto the machine, so it becomes very important to remove the taint as soon as the pod you're trying to move there schedules there, and you also have to remove the taint in the case where that pod actually doesn't move (e.g. rescheduler didn't know about some scheduler-specific predicate that actually makes it infeasible on that machine, so the pod ends up back where it started). So the "garbage collection" becomes much more time-sensitive and critical, versus the "hint" approach of the rate limiter. Also now that I think about it, you would need the controller to put the toleration on the replacement pod (because that pod doesn't exist at the time the rescheduler makes its request), so this only solves the problem of how to communicate the request from the controller to the scheduler, not how to communicate the request form the rescheduler to the controller. The rate limiter object (or node status thing) allows the rescheduler to write the info and scheduler to consume it, without the controller having to get involved.

davidopp commented:
> Yet another option is to create another polymorphic controller subresource similar to /scale, say /move (which is also conveniently both a noun and verb), which could be invoked on the responsible controller identified by controllerRef to request the pod in question to be moved gracefully by the controller.

This would be useful for rescheduler but I think it doesn't handle the other use cases for the rate lmiter/avoid machine, like backing off when a kubelet admission rejection loop on a machine is detected (kubelet has more information than scheduler), or when a "pod of death" relative to a specific machine is detected.

> That controller could generate a one-time anti-affinity constraint on the replacement pod.

I considered the one-time anti-affinity idea but there are two reasons I don't think it's a good idea:

1) In the rescheduler case (though maybe not the other cases of the rate limiter -- I'm not sure) I think we want the semantics "avoid this machine unless it is the only feasible machine." This is not captured by hard affinity (for obvious reasons) nor by soft affinity (because the semantics I described implies we want soft node affinity to be the highest-weighted priority function when processing this request, but there's no way to guarantee the scheduler is configured that way). 

2) It is confusing to the user to see node affinities magically added behind their back. This also makes updates, etc. more complicated (though I know things like patch are intended to handle this).
 
Assuming we don't use nodeAffinity for the reasons I mentioned above, then the subresource helps with half of the problem (how rescheduler tells the controller) but not the other half (how controller tells the scheduler). Whereas using a new rate limiter API object (or equivalent information in the NodeSpec or NodeStatus) collapses this to one operation: rescheduler writes `<controllerRef, Node>` to the object or `<controllerRef>` to the NodeSpec/NodeStatus, and you're done.

bgrant0607 commented:
I'm not convinced the rescheduler should be responsible for the case of a pod not scheduling. The scheduler responsible for that pod knows whether that is the case, and either something can/should be preempted to make it schedule, or not.

davidopp commented:
> I'm not convinced the rescheduler should be responsible for the case of a pod not scheduling. The scheduler responsible for that pod knows whether that is the case, and either something can/should be preempted to make it schedule, or not.

Can you clarify which of my comments this was a reply to?

davidopp commented:
Oh and BTW - no, this is not for 1.2.

davidopp commented:
Somewhere I thought I saw @bgrant0607 make a comment about namespace scopes somewhere. I can't find it anymore, so I'll just give a reply to a possibly-imagined question here:

To the extent that these prefer/avoid rules refer to controllers, which live in namespaces, the rules are naturally scoped. But if we store the information in the NodeStatus, the information on one node will naturally contain rules for controllers from multiple namespaces, since nodes are unscoped. The only way we could store the information partitioned by namespace is if we use the rate limiter object approach (not stored in nodes) and have one object per namespace, that covers the controllers in that namespace.

In general, schedulers are not tied to namespaces. As an obvious example, the default scheduler schedules for all namespaces. So a scheduler would need to pay attention to the rules that reference controllers from all namespaces regardless of how it is implemented. Also, we are not running one rescheduler per namespace (by definition that couldn't work), so it needs to create rules across all namespaces.

I have no idea if that random collection of facts addresses the (possibly imagined) question.

bgrant0607 commented:
@davidopp My comment about the case of a pod not scheduling refers to the need for a "prefer" mechanism.

bgrant0607 commented:
Continuing discussion on #12140

davidopp commented:
> Continuing discussion on #12140

At this point I think that's too late. I think it's better to have all of the discussion in one place, albeit the wrong place, than half the wrong place and half the right place. We can put a pointer here (besides the auto-generated one) from the issue. Anyway this discussion is diverging, not converging, so it's not like we're going to have a nice concise summary in the issue any time soon.

davidopp commented:
Never mind, I see you went back to the rescheduler issue, not the api.DeleteOptions issue. That makes sense, for discussion of rescheduler in general. I do think if we're going to discuss the specific issue of how to store/transmit directives from rescheduler to controllers and/or schedulers, we should do it here, since there is so much thread here already on that topic.

bgrant0607 commented:
Since we're not going to be able to keep pods around after deletion in the near future, the idea of a "rate limiter" would be useful. Associating it with nodes would scale better than with controllers. It would have to be bounded in size. I'd be ok with putting it in NodeStatus. We need to make NodeStatus updates cheaper, anyway (#14733).

However, I wouldn't call it "RaterLimiter". Something like "RecentEvictions" would be more appropriate. It should contain a (bounded-size) list of "PodSignature", Reason, Message, and EvictionTime. 

However, I'm concerned about recording PII in NodeStatus. This is a problem with recording image names in NodeStatus, also (https://github.com/kubernetes/kubernetes/pull/21182#issuecomment-186264570). We'd need to think about how to address that problem, or would need to put such scheduler hints elsewhere.

jiangyaoguo commented:
I would prefer to "RecentEvictions" because "RateLimiter" has been used widely in kubernetes for limiting http requests. For garbage collecting, if we use EvictionTime, we may need to set a unified TTL for all "PodSignature". "SuggestDeleteAt" is more flexible since it is user defined.

davidopp commented:
@bgrant0607 and I discussed this earlier this week. I will write something up in the next couple of days. I apologize for how long this is dragging out, but I think everyone was right that my initial idea described here was not the best solution.

davidopp commented:
I will write up something more complete/general regarding rescheduler later, but here is our revised proposal for this particular issue:

Add to `NodeStatus` a field

```
// Bounded-sized list of signatures of pods that should avoid this node, sorted
// in timestamp order from oldest to newest. Size of the slice is unspecified.
PreferAvoidPods []PreferAvoidPodsEntry
```

defined as follows

```
type PreferAvoidPodsEntry struct {
   PodSignature PodSignature
   EvictionTime *unversioned.Time
   Reason string
   Message string
}

// Describes the class of pods that should avoid this node.
// Exactly one field should be set.
type PodSignature struct {
    ControllerRef *ControllerRef
}

type ControllerRef struct {
   Kind string
   Namespace string
   Name string 
   APIVersion string   // for API group   
}

```

PodSignature is a struct to allow for future expansion. Right now it only has ControllerRef.

cc/ @bgrant0607

jiangyaoguo commented:
@davidopp If there is no objection, I'll update this pr to Add PreferAvoidPods to NodeStatus.

davidopp commented:
Yes that's fine.

jiangyaoguo committed:
Use PreferAvoidPods annotation to avoid pods being scheduled to specific node.
1. define PreferAvoidPods annotation
2. add PreferAvoidPodsPriority
3. validate AvoidPods in node annotations

jiangyaoguo commented:
We have defined a general "ObjectReference" and implemented some operations in pkg/ref.go. So we can reuse it to reference controller or other objects.

bgrant0607 commented:
Please don't reuse ObjectReference. We're moving in the direction of more specialized types, so that unused fields won't be incorrectly persisted or documented, so we can provide context-specific documentation generated from the types. etc.

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+type PreferAvoidPodsEntry struct {
+	//  The class of pods.
+	PodSignature PodSignature `json:"podSignature"`
+	// The time at which this eviction happened.
```
To make this more general, say "Time at which this entry was added to the list"

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+	PodSignature PodSignature `json:"podSignature"`
+	// The time at which this eviction happened.
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason for this eviction.
```
To make this more general, say "(brief) reason this entry was added to the list."

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason for this eviction.
+	Reason string `json:"reason,omitempty"`
+	// Human readable message indicating details about this eviction.
```
To make this more general, say "Human-readable message indicating why this entry was added to the list."

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+// Describes the class of pods that should avoid this node.
+// Exactly one field should be set.
+type PodSignature struct {
+	// Reference to replication-controller of pods that should avoid this node.
```
To make this more general, just say "Reference to controller whose pods should avoid this node."

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+type ControllerRef struct {
+	// Kind of the referent.
+	Kind string `json:"kind,omitempty"`
+	// Namespace of the replication-controller.
```
s/replication-controller/controller/

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+	Kind string `json:"kind,omitempty"`
+	// Namespace of the replication-controller.
+	Namespace string `json:"namespace,omitempty"`
+	// Name of the replication-controller.
```
s/replication-controller/controller/

davidopp commented on pkg/api/types.go:
```diff
@@ -1716,6 +1716,40 @@ type NodeStatus struct {
+	Namespace string `json:"namespace,omitempty"`
+	// Name of the replication-controller.
+	Name string `json:"name,omitempty"`
+	// API version of the replication-controller.
```
s/replication-controller/controller/

davidopp commented:
@jiangyaoguo Very sorry for my long delay in returning to review this.

davidopp commented:
@jiangyaoguo please let me know when you have updated the PR

jiangyaoguo commented:
@davidopp I have updated the pr as suggested. Sorry for forgetting declaring and reply.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -2103,6 +2103,40 @@ type NodeStatus struct {
+
+// ControllerRef contains enough information to let you inspect or modify the referred controller.
+type ControllerRef struct {
+	// Kind of the referent.
```
Kind of the controller.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -2103,6 +2103,40 @@ type NodeStatus struct {
+	PreferAvoidPods []PreferAvoidPodsEntry `json:"preferAvoidPods,omitempty"`
+}
+
+// Describes prefer/avoid rules when scheduling pods to this node.
```
Describes a class of pods that should avoid this node.

jiangyaoguo replied:
@davidopp Doesn't PodSignature describes the class of pods that should avoid this node?

davidopp replied:
Yes. The problem is just an English ambiguity. The way it's currently written, someone might think it means "Describes rules that say a pod should prefer or avoid this node." I'm just suggesting to get rid of the word "prefer" so that people are clear it is just talking about avoiding the node. It would be fine to say "Describes avoid rules when scheduling pods to this node" -- but I think the way I wrote it is a little clearer.

The name PreferAvoidPods has the same problem (it really means PreferToAvoidPods, but some people might interpret it as PreferOrAvoidPods) but I don't think we need to rename it; it's sufficient to just make sure the comment is clear. (Your comment on PreferAvoidPods is fine; it's just the one on PreferAvoidPodsEntry that I thought might be confusing.)

jiangyaoguo replied:
Yes, the current comments on PreferAvoidPodsEntry is ambiguity. It should be more clear. What I want to say is “Describes a class of pods that should avoid this node" is duplicate with comments on PodSignature objects.

jiangyaoguo commented:
@k8s-bot test this please

jiangyaoguo commented:
@davidopp Updated PR as suggested.

davidopp commented:
@bgrant0607 can you PTAL at the API. Two questions:
1) Does this ControllerRef look right? I assume we will reuse it for the ControllerRef that points from pods to their controllers, when we eventually implement that?
2) Should PreferAvoidPods be an alpha annotation on Node, instead of a field in Node, for the first version?

wojtek-t commented on pkg/api/v1/types.go:
```diff
@@ -2108,7 +2108,41 @@ type NodeStatus struct {
+}
+
+// ControllerRef contains enough information to let you inspect or modify the referred controller.
+type ControllerRef struct {
```
My thoughts about ControllerRef:
- controllerRef in my opinion is supposed to be a generic concept for all objects (as an example, controllerRef of ReplicaSet can be set to deployment).
- we shouldn't reinvent the way of referencing from scratch - we already have "ObjectReference" object defined and we should use it here.

davidopp commented:
@jiangyaoguo Please make the following changes (1 and 2 below are based on discussion with @bgrant0607 ):

1) PreferAvoidPods should be an alpha annotation on Node, rather than a field in NodeStatus. You can look at @HaiyangDING's multi-scheduler PR (#17865) for an example of how annotations work. (There the annotation was on Pod, instead of Node.)

2) This "controller ref" will be different from the controllerRef that @wojtek-t is talking about. So we should rename it in your PR. I would suggest s/ControllerRef/PodController/.

3) I think you should also put in this PR the scheduler priority function that gives priority 10 to every node that does not have the controller of the new pod (pod you are scheduling) listed in PreferAvoidPods. (We may want to give that a higher weight than 1 but we can discuss that later.)

4) We need to make this API "invisible" until the API server part (that @mml is working on, that writes and maintains the PreferAvoidPods annotation) is finished. That will allow you to submit this PR before he is finished. @erictune knows how to do this, but I don't; I'll update this issue when I get that information from him (or he'll add it here).

jiangyaoguo commented:
@k8s-bot test this please

davidopp commented:
@jiangyaoguo Please run "hack/update-all.sh" and push the new commit. (That is why the test is failing.)

jiangyaoguo commented:
@davidopp It's strange for that it can't pass "verify" but "hack/update-all.sh" make no difference.

davidopp commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
 	Images []ContainerImage `json:"images,omitempty"`
 }
 
+// AvoidPods desribes pods that should avoid this node.
```
Add: "This is the value for a Node annotation with key scheduler.alpha.kubernetes.io/preferAvoidPods and will eventually become a field of NodeStatus."

Also, although the multi-scheduler annotation doesn't have validation, I think AvoidPods should, because it is fairly complex. (Multi-scheduler annotation value is just a string, whereas value here is a JSON object.) You can look at how @kevin-wangzefeng added validation for node affinity: trace the code starting at the call to ValidateAffinityInPodAnnotations() from ValidatePodSpecificAnnotations() in pkg/api/validation/validation.go (you will need to add a function ValidateNodeSpecificAnnotations() since one does not exist yet, and call it from ValidateNode(); you can just follow the pattern that was used for pod). I think the only thing you need to validate is that exactly one field of PodSignature is set, so this should be easy. (Obviously only zero or one fields of PodSignature can be set today, but it will be good to have this validation code later if we add more fields.)

davidopp commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+	PodSignature PodSignature `json:"podSignature"`
+	// Time at which this entry was added to the list.
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason for this entry was added to the list.
```
s/for/why/

davidopp commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+// Exactly one field should be set.
+type PodSignature struct {
+	// Reference to controller whose pods should avoid this node.
+	PodController *PodController `json:"podController"`
```
I think pointer type has to be "omitempty"

davidopp commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+	PodController *PodController `json:"podController"`
+}
+
+// PodController contains enough information to let you inspect or modify the referred controller.
```
@bgrant0607 This also needs group, right?

davidopp commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+}
+
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
```
I think none of these should be "omitempty" because they are all needed to resolve the name of a controller. Though I need to look into how the lister uses APIVersion.

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
 	return float64(requested) / float64(capacity)
 }
+
+type NodePreferAvoid struct {
```
s/NodePreferAvoid/NodePreferAvoidPods/

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+	controllerLister algorithm.ControllerLister
+}
+
+func NewNodePreferAvoidPriority(controllerLister algorithm.ControllerLister) algorithm.PriorityFunction {
```
s/NewNodePreferAvoidPriority/NewNodePreferAvoidPodsPriority/

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+	return nodePreferAvoid.CalculateNodePreferAvoidPriority
+}
+
+func (npa *NodePreferAvoid) CalculateNodePreferAvoidPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
```
s/CalculateNodePreferAvoidPriority/CalculateNodePreferAvoidPodsPriority/

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+
+	result := []schedulerapi.HostPriority{}
+
+	rcs, err := npa.controllerLister.GetPodControllers(pod)
```
GetPodControllers() only gives you ReplicationControllers. You should also do GetPodReplicaSets() (see how this is done in CalculateSpreadPriority() in plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go). Once we have controllerRef we can just use that, to find the pod's controller, so please also add

// TODO: Once we have controllerRef, use it to find controller for the pod.

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+		}
+		for _, avoid := range avoids.PreferAvoidPods {
+			for _, rc := range rcs {
+				if avoid.PodSignature.PodController.Name == rc.Name && avoid.PodSignature.PodController.Namespace == rc.Namespace {
```
You should also check kind. Since you know rc is a ReplicationController, you can just check that avoid.PodSignature.PodController.Kind == "ReplicationController". (Of course when you also add ReplicaSet you would do the same for the ReplicaSets.) I'm actually not sure what to do about version; I'll get back to you about that one later.

liggitt replied:
Need group-qualified kind, at least

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+			for _, rc := range rcs {
+				if avoid.PodSignature.PodController.Name == rc.Name && avoid.PodSignature.PodController.Namespace == rc.Namespace {
+					avoidNodes[node.Name] = true
+					break
```
this only breaks from the "range rcs" loop but you also want to break from the "range avoid.PreferAvoidPods" loop

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+		}
+	}
+
+	//If all the nodes prefer avoiding this pod, we still can schedule pod to these node.
```
This loop isn't necessary. Total priority is just sum over the output of all priority functions, so if a priority function gives the same score to all nodes, the priority function has no effect, and the other priority functions determine the "best" node.

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,74 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+		return result, nil
+	}
+
+	//score int - scale of 0-10
```
score is supposed to be 1 to 10 but in this case I think 0 is fine. I'll add to #24720 that we should make score be an abstract datatype so we can make sure it is used correctly.

davidopp commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -168,5 +168,14 @@ func defaultPriorities() sets.String {
 			},
 		),
+		factory.RegisterPriorityConfigFactory(
+			"NodePreferAvoidPriority",
```
s/NodePreferAvoidPriority/NodePreferAvoidPodPriority/

davidopp commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -168,5 +168,14 @@ func defaultPriorities() sets.String {
+			"NodePreferAvoidPriority",
+			factory.PriorityConfigFactory{
+				Function: func(args factory.PluginFactoryArgs) algorithm.PriorityFunction {
+					return priorities.NewNodePreferAvoidPriority(args.ControllerLister)
```
s/NewNodePreferAvoidPriority/NewNodePreferAvoidPodsPriority/

davidopp commented on plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go:
```diff
@@ -168,5 +168,14 @@ func defaultPriorities() sets.String {
+				Function: func(args factory.PluginFactoryArgs) algorithm.PriorityFunction {
+					return priorities.NewNodePreferAvoidPriority(args.ControllerLister)
+				},
+				Weight: 1,
```
You should give this a high weight, because when a pod's controller is in PreferAvoidPods, we want that node to be used only as a last choice. It's not easy to choose a good weight because it depends on the other weights (and in fact today the total priority can overflow, see #24720 which I just filed). I would suggest setting this to 10000 and adding this comment above it:
```
// Set this weight large enough to override all other priority functions.
// TODO: Figure out a better way to do this, maybe at same time as fixing #24720.
```

davidopp commented on pkg/api/helpers_test.go:
```diff
@@ -297,3 +297,74 @@ func TestGetAffinityFromPod(t *testing.T) {
+	}
+
+	for i, tc := range testCases {
+		_, err := GetAvoidPodsFromNodeAnnotations(tc.node.Annotations)
```
Can you also check the value you got back (in the case where err == nil) to make sure the fields are what you expect?

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities_test.go:
```diff
@@ -929,3 +930,142 @@ func TestPrioritiesRegistered(t *testing.T) {
+				},
+			},
+			expectedList: []schedulerapi.HostPriority{{"machine1", 10}, {"machine2", 10}},
+			test:         "pod should avoid these nodes, but there's no other available nodes, all these nodes get highest priority score",
```
this test (and the corresponding code) isn't necessary; see my comment in the priority function

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities_test.go:
```diff
@@ -929,3 +930,142 @@ func TestPrioritiesRegistered(t *testing.T) {
 	}
 }
+
+func TestNodePreferAvoidPriority(t *testing.T) {
```
Please also also test the case where pod (and annotation) has a ReplicaSet instead of a RepilcationController, once you add the replica set lister.

davidopp commented:
You can ignore my comment 4 here https://github.com/kubernetes/kubernetes/pull/20699#issuecomment-211643254
So I think this PR is fine once you make the changes I mentioned.

Thanks!

liggitt commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+	// Name of the controller.
+	Name string `json:"name,omitempty"`
+	// API version of the controller.
+	APIVersion string `json:"apiversion,omitempty"`
```
I think group and version should be specified individually

liggitt replied:
Especially if we expect to be able to continue to look up the reference in future versions

liggitt commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,44 @@ type NodeStatus struct {
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
+	// Kind of the controller.
+	Kind string `json:"kind,omitempty"`
```
Kind or Resource? Will discovery be used to resolve the lookup, or just the mechanical kindToResource?

jiangyaoguo commented:
@davidopp @liggitt Updated pr except group probelem in reference.  Is` group` necessary for controller reference? Or use a `GroupVersionKind`?

liggitt commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,46 @@ type NodeStatus struct {
+	// Name of the controller.
+	Name string `json:"name"`
+	// API version of the controller.
+	APIVersion string `json:"apiversion"`
```
can you describe what this would be set to for a pod created by a controller that is surfaced in multiple versions of the API?

liggitt replied:
also, `apiVersion`

liggitt commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,46 @@ type NodeStatus struct {
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
+	// Kind of the controller.
+	Kind string `json:"kind"`
```
how will this be used? should this be a Kind or a Resource?

liggitt commented on pkg/api/types.go:
```diff
@@ -1733,6 +1733,46 @@ type NodeStatus struct {
+}
+
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
```
how is this different from an ObjectReference?

liggitt replied:
should UID be on here as well?

davidopp commented:
Sorry for the delay in responding. I talked to @bgrant0607 about this yesterday and he felt that since controllerRef is going to be in 1.3, we should re-use that here instead of defining a new PodController type. (And I agree, this makes sense.)

It will make the scheduler code even simpler -- it just compares the controllerRef of the pod-you-are-scheduling to all of the controllerRef's in the PreferAvoidPods for the node.

Unfortunately, that means this PR is blocked on implementing controllerRef. I filed an issue for it (#24946). @gmarek is going to implement that.

So, please wait to finish this PR until controllerRef is implemented.

davidopp commented:
We decided not to create a new controllerRef type, instead we will use existing OwnerReference (which is already in the codebase).

So in this PR, you should change

```
type PodSignature struct {
	// Reference to controller whose pods should avoid this node.
	PodController *PodController `json:"podController,omitempty"`
}
```

to

```
type PodSignature struct {
	// Reference to controller whose pods should avoid this node.
	PodController *OwnerReference `json:"podController,omitempty"`
}
```

and delete the `PodController` type.

That should unblock this PR.

Thanks and sorry for the delay.

jiangyaoguo commented:
@k8s-bot  test this issue: #26138

jiangyaoguo commented:
@davidopp Updated with OwnerReference replacement and priority function change.

jiangyaoguo commented:
@davidopp Rebased.

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1661,6 +1661,27 @@ func ValidateNodeSelector(nodeSelector *api.NodeSelector, fldPath *field.Path) f
+	if len(avoids.PreferAvoidPods) != 0 {
+		for _, pa := range avoids.PreferAvoidPods {
+			if pa.PodSignature.PodController == nil {
+				allErrs = append(allErrs, field.Required(fldPath.Child("PodSignature"), ""))
```
We should wait until #25567 merges and then also verify that
```
if (pa.PodSignature.PodController != nil) {
   // verify that *(*(pa.PodSignature.PodController).Controller) == true
}
```

davidopp commented on pkg/api/validation/validation_test.go:
```diff
@@ -4055,6 +4055,42 @@ func TestValidateNode(t *testing.T) {
+							    "preferAvoidPods": [
+							        {
+							            "podSignature": {
+							                "podController": {
```
where did namespace come from? I don't see that in OwnerReference.
Also, once #25567 merges, you should add controller field (it's a bool)

davidopp commented on pkg/api/validation/validation_test.go:
```diff
@@ -4444,6 +4507,57 @@ func TestValidateNodeUpdate(t *testing.T) {
+							    "preferAvoidPods": [
+							        {
+							            "podSignature": {
+							                "podController": {
```
see previous comment

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -273,3 +273,79 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+
+	result := []schedulerapi.HostPriority{}
+
+	// TODO: Once we have controllerRef, use it to find controller for the pod.
```
s/controllerRef/ownerReference fully implemented/

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities_test.go:
```diff
@@ -929,3 +930,135 @@ func TestPrioritiesRegistered(t *testing.T) {
+							                    "apiVersion": "v1",
+							                    "kind": "ReplicaSet",
+							                    "name": "foo",
+												"uid": "qwert12345"
```
indentation

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/priorities_test.go:
```diff
@@ -929,3 +930,135 @@ func TestPrioritiesRegistered(t *testing.T) {
+						                        "apiVersion": "v1",
+							                    "kind": "ReplicationController",
+							                    "name": "foo",
+												"uid": "abcdef123456"
```
indentation

davidopp commented:
Thanks for updating. Wait until #25567 merges and then make the changes I described, and then I'll LGTM it.

jiangyaoguo commented:
@davidopp Update pr and #25567 merges now.

davidopp commented on pkg/api/helpers_test.go:
```diff
@@ -295,3 +295,98 @@ func TestGetAffinityFromPod(t *testing.T) {
+							                "podController": {
+							                    "kind": "ReplicationController",
+							                    "apiVersion": "v1"
+							            "reason": "some reason",
```
indentation problem here

davidopp commented on pkg/api/helpers_test.go:
```diff
@@ -295,3 +295,98 @@ func TestGetAffinityFromPod(t *testing.T) {
+							                    "kind": "ReplicationController",
+							                    "apiVersion": "v1"
+							            "reason": "some reason",
+							            "message": "some message"
```
indentation problem here

jiangyaoguo replied:
This case test a incorrect Json format with no end symbol of struct "podController" and "podSignature". The indentation of "reason" and "message" is correct.

davidopp commented on pkg/api/helpers_test.go:
```diff
@@ -295,3 +295,98 @@ func TestGetAffinityFromPod(t *testing.T) {
+			expectErr: false,
+		},
+		{
+			node: &Node{
```
add a comment why you expect this to fail

davidopp commented:
Thanks! Just a few comments, which are all trivial. Once you fix them I will LGTM.

wojtek-t commented:
I don't understand title of this PR - this doesn't touch NodeStatus at all.

davidopp commented:
Thanks for noticing, the title is out-of-date (we decided to use annotation for first version). I will update it now.

jiangyaoguo commented:
@davidopp updated.

davidopp commented on pkg/api/v1/types.go:
```diff
@@ -2392,6 +2392,34 @@ type NodeStatus struct {
 
 type UniqueDeviceName string
 
+// AvoidPods desribes pods that should avoid this node. This is the value for a
```
s/desribes/describes/

davidopp commented on pkg/api/types.go:
```diff
@@ -1989,6 +1989,34 @@ type NodeStatus struct {
 
 type UniqueDeviceName string
 
+// AvoidPods desribes pods that should avoid this node. This is the value for a
```
s/desribes/describes/

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1698,6 +1698,39 @@ func ValidateNodeSelector(nodeSelector *api.NodeSelector, fldPath *field.Path) f
+	return allErrs
+}
+
+// validateTaints tests if given taints have valid data.
```
s/validateTaints/validatePreferAvoidPodsEntry/

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1698,6 +1698,39 @@ func ValidateNodeSelector(nodeSelector *api.NodeSelector, fldPath *field.Path) f
+		allErrors = append(allErrors, field.Required(fldPath.Child("PodSignature"), ""))
+	} else {
+		if *(avoidPodEntry.PodSignature.PodController.Controller) != true {
+			allErrors = append(allErrors, field.Invalid(fldPath.Child("PodSignature").Child("PodController").Child("Controller"), *(avoidPodEntry.PodSignature.PodController.Controller), "must be true when ownerReferrence points to a controller"))
```
The error here should be a little different -- something like
`allErrors = append(allErrors, field.Invalid(fldPath.Child("PodSignature").Child("PodController"), avoidPodEntry.PodSignature.PodController), "must point to a controller"))`

jiangyaoguo commented:
@davidopp updated.

jiangyaoguo commented:
@davidopp rebased.

davidopp commented:
LGTM

Very sorry for the delay in reviewing. Please rebase one more time, and then I will add LGTM label.

Thanks for your work on this!

davidopp commented:
Please squash the commits to one commit.

davidopp commented:
@jiangyaoguo your new commit has a compiler error:

```
# k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/priorities
plugin/pkg/scheduler/algorithm/priorities/priorities.go:307: nodes.Items undefined (type []*"k8s.io/kubernetes/pkg/api".Node has no field or method Items)
plugin/pkg/scheduler/algorithm/priorities/priorities.go:314: nodes.Items undefined (type []*"k8s.io/kubernetes/pkg/api".Node has no field or method Items)
```

I didn't look carefully, but it might be that you need to update your code to incorporate the change made in #28781

jiangyaoguo commented:
@davidopp  I encounter some network problem now.  Will update soon.

jiangyaoguo commented:
@davidopp

wojtek-t commented on plugin/pkg/scheduler/algorithm/priorities/priorities.go:
```diff
@@ -277,3 +277,79 @@ func fractionOfCapacity(requested, capacity int64) float64 {
+	return nodePreferAvoid.CalculateNodePreferAvoidPodsPriority
+}
+
+func (npa *NodePreferAvoidPod) CalculateNodePreferAvoidPodsPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
```
@davidopp @jiangyaoguo - do we know the preformance impact of it? Since you are switching it on by default, it would be good to know...

wojtek-t replied:
So let's do this way:
1) let's merge as it is now (basically in selector spreading we are still calling all the functions that we have here, so it shouldn't be worse)
2) Once this is merged, I will do a pass over it and fix some issues that I already fixed for other priorities/predicates
3) once ControllerRef is ready, I will switch to that (together with selector_spreading)

[Generally, the asymptotical complexity of it is the same as selector spreading, We should probably solve both of them in the same way in the future.]

davidopp replied:
SGTM. Thanks!

davidopp commented:
LGTM

vishh commented:
@davidopp Is this feature documented anywhere?
